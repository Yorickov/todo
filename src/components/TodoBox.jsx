import React from 'react';
import NewTaskForm from './NewTaskForm.jsx';
import Tasks from './Tasks.jsx';
import Filter from './Filter.jsx';

const TodoBox = () => (
  <div className="row">
    <div className="col-6">
      <NewTaskForm />
      <Filter />
      <Tasks />
    </div>
  </div>
);
export default TodoBox;
