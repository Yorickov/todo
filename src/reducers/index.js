import { createReducer, combineReducers } from '@reduxjs/toolkit';
import _ from 'lodash';
import {
  updateNewTaskText,
  addTask,
  removeTask,
  toggleTaskState,
  inverseTaskTheme,
  setTasksFilter,
} from '../actions';

const text = createReducer('', {
  [updateNewTaskText]: (_state, { payload: { text } }) => text,
  [addTask]: () => '',
});

const tasks = createReducer({ byId: {}, allIds: [], currentFilterName: 'all' }, {
  [addTask]: (state, { payload: { task } }) => {
    const { byId, allIds } = state;
    return {
      ...state,
      byId: { ...byId, [task.id]: task },
      allIds: [...allIds, task.id],
    };
  },
  [removeTask]: (state, { payload: { id } }) => {
    const { byId, allIds } = state;
    return {
      ...state,
      byId: _.omit(byId, id),
      allIds: _.without(allIds, id),
    };
  },
  [toggleTaskState]: (state, { payload: { id } }) => {
    const { byId } = state;
    const task = byId[id];
    const newState = task.state === 'active' ? 'finished' : 'active';
    const updatedTask = { ...task, state: newState };
    return {
      ...state,
      byId: { ...byId, [task.id]: updatedTask },
    };
  },
  [setTasksFilter]: (state, { payload: { filterName } }) => (
    { ...state, currentFilterName: filterName }),
});

const tasksUIState = createReducer({}, {
  [addTask]: (state, { payload: { task } }) => ({
    ...state,
    [task.id]: { theme: 'light' },
  }),
  [inverseTaskTheme]: (state, { payload: { task } }) => {
    const currentTheme = state[task.id].theme;
    const mapping = { white: 'light', light: 'white' };
    return { ...state, [task.id]: { theme: mapping[currentTheme] } };
  },
});

export default combineReducers({
  text,
  tasks,
  tasksUIState,
});
