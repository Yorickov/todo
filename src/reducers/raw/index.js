import { createReducer, combineReducers } from '@reduxjs/toolkit';

const increment = createReducer(0, {
  INCREMENT: (state) => state + 1,
});

const text = createReducer('', {
  TEXT_UPDATE: (_state, action) => action.payload.text,
  TEXT_RESET: () => '',
});

export default combineReducers({
  increment,
  text,
});
