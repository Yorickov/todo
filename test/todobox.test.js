import React from 'react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import reducers from '../src/reducers';
import TodoBox from '../src/components/TodoBox.jsx';

it('Store', () => {
  const store = configureStore({
    reducer: reducers,
  });

  const vdom = (
    <Provider store={store}>
      <TodoBox />
    </Provider>
  );

  const wrapper = mount(vdom);
  expect(wrapper.render()).toMatchSnapshot();

  const newTaskInput = wrapper.find('input[type="text"]');
  const newTaskSubmit = wrapper.find('button[type="submit"]');

  newTaskInput.simulate('change', { target: { value: 'first task' } });
  expect(wrapper.render()).toMatchSnapshot();

  newTaskSubmit.simulate('submit');
  expect(wrapper.render()).toMatchSnapshot();

  newTaskInput.simulate('change', { target: { value: 'another task' } });
  expect(wrapper.render()).toMatchSnapshot();

  newTaskSubmit.simulate('submit');
  expect(wrapper.render()).toMatchSnapshot();
  expect(wrapper.find('.list-group-item')).toHaveLength(2);

  wrapper
    .find('[data-test="task-change-theme"]').first()
    .simulate('click');
  expect(wrapper.find('.list-group-item').first().hasClass('bg-white')).toBeTruthy();

  wrapper
    .find('[data-test="task-change-theme"]').first()
    .simulate('click');
  expect(wrapper.find('.list-group-item').first().hasClass('bg-light')).toBeTruthy();

  wrapper
    .find('[data-test="task-toggle-state"]').first()
    .simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  const activeFilterButton = wrapper.find('[data-test="task-filter-active"]');
  const finishedFilterButton = wrapper.find('[data-test="task-filter-finished"]');

  activeFilterButton.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
  expect(wrapper.find('.list-group-item')).toHaveLength(1);

  finishedFilterButton.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
  expect(wrapper.find('.list-group-item')).toHaveLength(1);

  const allFilterButton = wrapper.find('[data-test="task-filter-all"]');
  allFilterButton.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
  expect(wrapper.find('.list-group-item')).toHaveLength(2);

  wrapper.find('.close').last()
    .simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
  expect(wrapper.find('.list-group-item')).toHaveLength(1);
});
