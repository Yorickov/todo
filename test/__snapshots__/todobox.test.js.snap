// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store 1`] = `
<div
  class="row"
>
  <div
    class="col-6"
  >
    <form
      action=""
      class="row"
    >
      <div
        class="col-10"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary col-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mt-3 d-flex justify-content-around"
    >
      All Tasks
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-active"
        type="button"
      >
        Active Tasks
      </button>
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-finished"
        type="button"
      >
        Finished Tasks
      </button>
    </div>
  </div>
</div>
`;

exports[`Store 2`] = `
<div
  class="row"
>
  <div
    class="col-6"
  >
    <form
      action=""
      class="row"
    >
      <div
        class="col-10"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value="first task"
        />
      </div>
      <button
        class="btn btn-primary col-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mt-3 d-flex justify-content-around"
    >
      All Tasks
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-active"
        type="button"
      >
        Active Tasks
      </button>
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-finished"
        type="button"
      >
        Finished Tasks
      </button>
    </div>
  </div>
</div>
`;

exports[`Store 3`] = `
<div
  class="row"
>
  <div
    class="col-6"
  >
    <form
      action=""
      class="row"
    >
      <div
        class="col-10"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary col-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mt-3 d-flex justify-content-around"
    >
      All Tasks
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-active"
        type="button"
      >
        Active Tasks
      </button>
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-finished"
        type="button"
      >
        Finished Tasks
      </button>
    </div>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex bg-light"
        >
          <button
            class="btn btn-secondary btn-sm me-3"
            data-test="task-change-theme"
            type="button"
          >
            <span>
              Theme
            </span>
          </button>
          <span
            class="me-auto d-flex align-items-center"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              first task
            </a>
          </span>
          <button
            class="close btn"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 4`] = `
<div
  class="row"
>
  <div
    class="col-6"
  >
    <form
      action=""
      class="row"
    >
      <div
        class="col-10"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value="another task"
        />
      </div>
      <button
        class="btn btn-primary col-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mt-3 d-flex justify-content-around"
    >
      All Tasks
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-active"
        type="button"
      >
        Active Tasks
      </button>
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-finished"
        type="button"
      >
        Finished Tasks
      </button>
    </div>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex bg-light"
        >
          <button
            class="btn btn-secondary btn-sm me-3"
            data-test="task-change-theme"
            type="button"
          >
            <span>
              Theme
            </span>
          </button>
          <span
            class="me-auto d-flex align-items-center"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              first task
            </a>
          </span>
          <button
            class="close btn"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 5`] = `
<div
  class="row"
>
  <div
    class="col-6"
  >
    <form
      action=""
      class="row"
    >
      <div
        class="col-10"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary col-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mt-3 d-flex justify-content-around"
    >
      All Tasks
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-active"
        type="button"
      >
        Active Tasks
      </button>
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-finished"
        type="button"
      >
        Finished Tasks
      </button>
    </div>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex bg-light"
        >
          <button
            class="btn btn-secondary btn-sm me-3"
            data-test="task-change-theme"
            type="button"
          >
            <span>
              Theme
            </span>
          </button>
          <span
            class="me-auto d-flex align-items-center"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              first task
            </a>
          </span>
          <button
            class="close btn"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
        <li
          class="list-group-item d-flex bg-light"
        >
          <button
            class="btn btn-secondary btn-sm me-3"
            data-test="task-change-theme"
            type="button"
          >
            <span>
              Theme
            </span>
          </button>
          <span
            class="me-auto d-flex align-items-center"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              another task
            </a>
          </span>
          <button
            class="close btn"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 6`] = `
<div
  class="row"
>
  <div
    class="col-6"
  >
    <form
      action=""
      class="row"
    >
      <div
        class="col-10"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary col-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mt-3 d-flex justify-content-around"
    >
      All Tasks
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-active"
        type="button"
      >
        Active Tasks
      </button>
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-finished"
        type="button"
      >
        Finished Tasks
      </button>
    </div>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex bg-light"
        >
          <button
            class="btn btn-secondary btn-sm me-3"
            data-test="task-change-theme"
            type="button"
          >
            <span>
              Theme
            </span>
          </button>
          <span
            class="me-auto d-flex align-items-center"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              <s>
                first task
              </s>
            </a>
          </span>
          <button
            class="close btn"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
        <li
          class="list-group-item d-flex bg-light"
        >
          <button
            class="btn btn-secondary btn-sm me-3"
            data-test="task-change-theme"
            type="button"
          >
            <span>
              Theme
            </span>
          </button>
          <span
            class="me-auto d-flex align-items-center"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              another task
            </a>
          </span>
          <button
            class="close btn"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 7`] = `
<div
  class="row"
>
  <div
    class="col-6"
  >
    <form
      action=""
      class="row"
    >
      <div
        class="col-10"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary col-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mt-3 d-flex justify-content-around"
    >
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-all"
        type="button"
      >
        All Tasks
      </button>
      Active Tasks
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-finished"
        type="button"
      >
        Finished Tasks
      </button>
    </div>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex bg-light"
        >
          <button
            class="btn btn-secondary btn-sm me-3"
            data-test="task-change-theme"
            type="button"
          >
            <span>
              Theme
            </span>
          </button>
          <span
            class="me-auto d-flex align-items-center"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              another task
            </a>
          </span>
          <button
            class="close btn"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 8`] = `
<div
  class="row"
>
  <div
    class="col-6"
  >
    <form
      action=""
      class="row"
    >
      <div
        class="col-10"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary col-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mt-3 d-flex justify-content-around"
    >
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-all"
        type="button"
      >
        All Tasks
      </button>
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-active"
        type="button"
      >
        Active Tasks
      </button>
      Finished Tasks
    </div>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex bg-light"
        >
          <button
            class="btn btn-secondary btn-sm me-3"
            data-test="task-change-theme"
            type="button"
          >
            <span>
              Theme
            </span>
          </button>
          <span
            class="me-auto d-flex align-items-center"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              <s>
                first task
              </s>
            </a>
          </span>
          <button
            class="close btn"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 9`] = `
<div
  class="row"
>
  <div
    class="col-6"
  >
    <form
      action=""
      class="row"
    >
      <div
        class="col-10"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary col-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mt-3 d-flex justify-content-around"
    >
      All Tasks
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-active"
        type="button"
      >
        Active Tasks
      </button>
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-finished"
        type="button"
      >
        Finished Tasks
      </button>
    </div>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex bg-light"
        >
          <button
            class="btn btn-secondary btn-sm me-3"
            data-test="task-change-theme"
            type="button"
          >
            <span>
              Theme
            </span>
          </button>
          <span
            class="me-auto d-flex align-items-center"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              <s>
                first task
              </s>
            </a>
          </span>
          <button
            class="close btn"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
        <li
          class="list-group-item d-flex bg-light"
        >
          <button
            class="btn btn-secondary btn-sm me-3"
            data-test="task-change-theme"
            type="button"
          >
            <span>
              Theme
            </span>
          </button>
          <span
            class="me-auto d-flex align-items-center"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              another task
            </a>
          </span>
          <button
            class="close btn"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Store 10`] = `
<div
  class="row"
>
  <div
    class="col-6"
  >
    <form
      action=""
      class="row"
    >
      <div
        class="col-10"
      >
        <input
          class="form-control mr-3"
          placeholder="I am going..."
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="btn btn-primary col-2"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="mt-3 d-flex justify-content-around"
    >
      All Tasks
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-active"
        type="button"
      >
        Active Tasks
      </button>
      <button
        class="btn btn-link border-0 p-0"
        data-test="task-filter-finished"
        type="button"
      >
        Finished Tasks
      </button>
    </div>
    <div
      class="mt-3"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex bg-light"
        >
          <button
            class="btn btn-secondary btn-sm me-3"
            data-test="task-change-theme"
            type="button"
          >
            <span>
              Theme
            </span>
          </button>
          <span
            class="me-auto d-flex align-items-center"
          >
            <a
              data-test="task-toggle-state"
              href="#"
            >
              <s>
                first task
              </s>
            </a>
          </span>
          <button
            class="close btn"
            type="button"
          >
            <span>
              ×
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
